### PR TITLE
Various additions

### PR DESCRIPTION
**Pull request by jodersky**
Currently WIP, these changes aim to bring the upickle community build up to feature parity with scala 2.

Some notable changes:

- add `macroR`, `macroW` and `macroRW`
- allow overriding field names through the `upickle.implicits.key` annotation
- allow unknown keys in reader
- handle missing keys in reader

Many tests still fail and/or need to be reactivated

### CODE REVIEW

File: upickle/test/src/upickle/MacroTests.scala [hunk 1]
```diff
@@ -61,8 +62,13 @@ object MacroTests extends TestSuite {
   val tests = Tests {
     test("mixedIn"){
       import MixedIn._
-      test - rw(Obj.ClsB(1), """{"i":1}""")
-      test - rw(Obj.ClsA("omg"), """{"s":"omg"}""")
+
+
+      println(upickle.default.read[Obj.ClsB]("""{"i":1}"""))
+      println(upickle.default.write(Obj.ClsB(1)))
```
Comment (jodersky) [inline upickle/test/src/upickle/MacroTests.scala]: currently debugging MacroTests

File: implicits/src/upickle/implicits/MacrosCommon.scala [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+// Common things for derivation
+trait MacrosCommon {
+
+  def serializeDefaults: Boolean = false
+
+  def objectAttributeKeyReadMap(s: CharSequence): CharSequence = s
+  def objectAttributeKeyWriteMap(s: CharSequence): CharSequence = s
+
+  def objectTypeKeyReadMap(s: CharSequence): CharSequence = s
+  def objectTypeKeyWriteMap(s: CharSequence): CharSequence = s
```
Comment (jodersky) [inline implicits/src/upickle/implicits/MacrosCommon.scala:12]: this stuff was moved from Api

File: implicits/src-0/upickle/implicits/macros.scala [hunk 3]
```diff
@@ -8,35 +8,59 @@ def getDefaultParmasImpl[T](using qctx: QuoteContext,
...
+      Expr.ofList(idents.map(_.seal.cast[Any]))
+
+    '{ $namesExpr.zip($identsExpr.map(_.asInstanceOf[AnyRef])).toMap }
+  } else {
+    '{ Map.empty }
+  }
 end getDefaultParmasImpl
 
-inline def summonAll[T <: Tuple]: List[_] =
+inline def summonList[T <: Tuple]: List[_] =
```
Comment (anatoliykmetyuk) [inline implicits/src-0/upickle/implicits/macros.scala:34]: summonAll was defined here temporarily while we didn't have one in Dotty. Now we do. Can we completely remove summonAll and summonList from here and use the Dotty implementation?

**Review by anatoliykmetyuk — APPROVED** [review]
Otherwise LGTM. Are the Dotty and Scala 2 tests passing locally?

**Pull request merged**